### PR TITLE
Structure_Engine: Transform methods added for structural elements

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Contributes to resolving #2322

### Test files

On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FStructure%5FEngine%2F%232322%2DTransform%2Egh&parent=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FStructure%5FEngine). I have disabled the test of FEMeshes because for some weird reason they slow GH down badly even if a completely unrelated part of the script gets modified - I could not come up with an explanation for that, maybe the reviewer will have an idea 😃 However, this is outside of the scope of this PR, because the above issue does not affect the correctness of the results, so please do not be afraid of enabling the disabled part.

### Changelog

- `Transform` methods added for `Node`, `Bar`, `RigidLink`, `Edge`, `Opening`, `Panel` and `FEMesh`

### Additional comments

I believe this is the most problematic/sensitive of all transform-related PRs so far, so probably worth @IsakNaslundBh's time for a proper review. I hope the provided test files will make the process a bit less painful.

### CODE REVIEW

**Comment by pawelbaran** [comment]
Tolerance and UT added @LMarkowski @IsakNaslundBh, the PR is ready to review

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:162]:
```suggestion
            result.ExternalEdges = result.ExternalEdges.Select(x => x.Transform(transform, tolerance)).ToList();
```

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:163]:
```suggestion
            result.Openings = result.Openings.Select(x => x.Transform(transform, tolerance)).ToList();
```

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:142]:
```suggestion
            result.Edges = result.Edges.Select(x => x.Transform(transform, tolerance)).ToList();
```

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:76]:
```suggestion
            result.StartNode = result.StartNode.Transform(transform, tolerance);
            result.EndNode = result.EndNode.Transform(transform, tolerance);
```

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:102]:
```suggestion
            result.PrimaryNode = result.PrimaryNode.Transform(transform, tolerance);
            result.SecondaryNodes = result.SecondaryNodes.Select(x => x.Transform(transform, tolerance)).ToList();
```

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:187]:
```suggestion
            result.Nodes = result.Nodes.Select(x => x.Transform(transform, tolerance)).ToList();
```

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Generally working great!

Just a comment on passing on the tolerances

**Comment by pawelbaran** [comment]
I have committed suggestions made by @IsakNaslundBh plus updated the GH script as well as the UTs including the extra test content provided by @IsakNaslundBh too. Should be ready for re-review now 👍

File: Structure_Engine/Modify/Transform.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/Transform.cs:55]: Another one (sorry).

Can we please null check the orientation property before transforming it?

PR open to default this property to null (https://github.com/BHoM/BHoM/pull/1184) so this would start to crash after that han been merged for non explicitly set orientation for nodes (and thereby start crashing for Bars and RigidLinks as well)

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Another quick one, see below

File: Structure_Engine/Modify/Transform.cs
Comment (pawelbaran) [reply Structure_Engine/Modify/Transform.cs]: Done :)

File: Structure_Engine/Modify/Transform.cs
Comment (LMarkowski) [inline Structure_Engine/Modify/Transform.cs:61]:
```suggestion
        [Description("Transforms the Bar's nodes and orientation by the transform matrix. Only rigid body transformations are supported.")]
```
Am I right?

**Review by LMarkowski — DISMISSED** [review]
Works and looks good to me! As in the physical added the erroneous cases to the UT.

**Comment by LMarkowski** [comment]
@BHoMBot check unit tests when you'll be upside up again.

**Comment by LMarkowski** [comment]
@BHoMBot check unit

**Comment by LMarkowski** [comment]
One more thought that I forgot to include in the review. The test only includes a validation for `Panel`s and is missing `Edge`s and `Openings`. Do the two latter methods occur outside the first one? If not, why keep them public? And if they are, is test for only `Panel` enough for us? For me - it is. But in the future I can imagine those two being 1 missing percent while doing analytics for unit test coverage across the engine.
@rolyhudson @al-fisher what are your thoughts on that foundational-side?

**Comment by pawelbaran** [comment]
@LMarkowski the methods for `Edge` and `Opening` are public to reflect the general philosophy of the BHoM (if something does not strongly need to be private, keep it public). I have not added type-specific tests for them (same as for `CeilingTile`) because they are being tested in the `Panel` transforms (and `Ceiling`, respectively).

Not sure if adding those tests as separate entities would bring any value - but can do, if you like, this should not be much of an effort.

**Comment by LMarkowski** [comment]
@pawelbaran as said I think test for panels is enough for me as well validation-wise. My only concern are missing UTs. They could be done even by transforming exploded panels. The question is if it's worth it.

**Comment by FraserGreenroyd** [comment]
> @pawelbaran as said I think test for panels is enough for me as well validation-wise. My only concern are missing UTs. They could be done even by transforming exploded panels. The question is if it's worth it.

It's always worth it, but it might not be valid to bring it in in this PR. The question is more accurately as whether it's worth it for this PR, not worth it overall.

**Comment by pawelbaran** [comment]
If it is worth it overall, I will add the UT's in right now - exploding the panels and kicking them in will take a minute, literally.

**Comment by LMarkowski** [comment]
> The question is more accurately as whether it's worth it for this PR, not worth it overall.

@FraserGreenroyd Well, having some time put in adding missing tests I'd say we should enforce adding the UT every time we add a new public method. Reduce backlog.

(What about private ones though 🤔)

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd Well, having some time put in adding missing tests I'd say we should enforce adding the UT every time we add a new public method. Reduce backlog.

Yup, agreed. That's why Environments/MEP are enforcing new Unit Tests on PRs (@pawelbaran found that when I requested changes on his MEP version of this PR solely for a missing unit test!). I believe structures are doing the same but @IsakNaslundBh can confirm that. I definitely recommend it though!

> (What about private ones though 🤔)

Private methods don't get exposed but are used by public methods to aid calculations, so the unit tests of public methods will be good enough (for now, we might change that later mind).

**Comment by pawelbaran** [comment]
Oke, I have added the missing UTs and updated the GH script.

**Comment by pawelbaran** [comment]
@BHoMBot check all

**Review by LMarkowski — APPROVED** [review]
Now it's perfect! Thank you.
RTM as soon as we have an approval from @BHoMBot

**Comment by pawelbaran** [comment]
> RTM as soon as we have an approval from @BHoMBot

You mean never?

image

**Comment by pawelbaran** [comment]
@BHoMBot you fool

**Comment by pawelbaran** [comment]
@BHoMBot check installer

**Comment by FraserGreenroyd** [comment]
> You mean never?
> 
> image

https://media.giphy.com/media/i0xGuo4o5PutVo0imJ/giphy.gif

**Comment by FraserGreenroyd** [comment]
Ok, bot issues aside, this is now ready for @IsakNaslundBh to give a decision from his changes requested and go for merge at that point :smile:

**Review by IsakNaslundBh — APPROVED** [review]
Have not tested after the changes, but trust @LMarkowski s test!

Issues I have raised have been resolved!

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**
